File: leetcode-js/105-construct-binary-tree-from-preorder-and-inorder-traversal.js
function TreeNode(val) {
  this.val = val;
  this.left = this.right = null;
}

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    if (!preorder.length) {
        return null;
    }

    return _buildTree(preorder, 0, preorder.length, inorder, 0, inorder.length);
};

function _buildTree(preorder, preStart, preEnd, inorder, inStart, inEnd) {
    'use strict';

    if (preStart === preEnd) {
        return null;
    }

    const val = preorder[preStart];
    const index = inorder.indexOf(val, inStart);
    const node = new TreeNode(val);

    node.left = _buildTree(preorder, preStart + 1, preStart + (index - inStart) + 1, inorder, inStart, index);
    node.right = _buildTree(preorder, preStart + (index - inStart) + 1, preEnd, inorder, index + 1, inEnd);

    return node;
}

console.log(buildTree([-1], [-1]))

File: leetcode-java/105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        TreeNode root = build(map, preorder, inorder, 0, preorder.length - 1, 0, inorder.length - 1);
        return root;
    }

    TreeNode build(Map<Integer, Integer> map, int[] preorder, int[] inorder, int preStart, int preEnd, int inStart, int inEnd) {
        if (preEnd - preStart < 0) {
            return null;
        }

        TreeNode node = new TreeNode(preorder[preStart]);

        int index = map.get(preorder[preStart]);
        int len = index - inStart;

        node.left = build(map, preorder, inorder, preStart + 1, preStart + len, inStart, inStart + len);
        node.right = build(map, preorder, inorder, preStart + len + 1, preEnd, index + 1, inEnd);

        return node;
    }
}
